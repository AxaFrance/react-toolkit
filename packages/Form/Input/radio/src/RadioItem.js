import React from 'react';
import PropTypes from 'prop-types';

import { InputList, InputConstants as Constants, withInput, omit } from '@axa-fr/react-toolkit-form-core';
import { InputManager } from '@axa-fr/react-toolkit-core';

const omitProperties = omit([
  'classModifier',
  'className',
  'isVisible',
  'helpMessage',
  'options'
]);

const RadioItem = ({
                     optionClassName,
                     disabled,
                     value,
                     isChecked,
                     id,
                     name,
                     inputRef,
                     readOnly,
                     onChange,
                     onBlur,
                     onFocus,
                     children,
                     label,
                     ...otherProps
                   }) => {
  const newLabel = children || label;
  const newId = InputManager.getInputId(id); // id is require on this component
  return (
    <div className={optionClassName} key={value}>
      <input
        className="af-form__input-radio"
        name={name}
        id={newId}
        type="radio"
        value={value}
        onChange={onChange}
        onBlur={onBlur}
        onFocus={onFocus}
        checked={isChecked}
        readOnly={readOnly}
        disabled={disabled}
        ref={inputRef}
        {...omitProperties(otherProps)}
      />
      <label className="af-form__label" htmlFor={newId}>
        <span className="af-form__description">{newLabel}</span>
      </label>
    </div>
  );
};

const propTypes = {
  ...Constants.propTypes,
  value: PropTypes.string,
  label: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
  isChecked: PropTypes.bool,
};
const defaultClassName = 'af-form__radio';
const defaultProps = {
  ...Constants.defaultProps,
  value: '',
  className: defaultClassName,
  classModifier: '',
  isChecked: false,
};

const props = ({ className, classModifier, disabled }) => ({
  optionClassName: InputList.getOptionClassName(
    className,
    classModifier,
    defaultClassName,
    disabled
  ),
});

const EnhancedComponent = withInput(
  defaultClassName,
  propTypes,
  defaultProps,
  {},
  props
)(RadioItem);
EnhancedComponent.displayName = RadioItem.name;

export default EnhancedComponent;
