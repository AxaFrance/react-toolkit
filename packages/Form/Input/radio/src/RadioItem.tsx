import React, { InputHTMLAttributes } from 'react';
import { InputManager, ClassManager } from '@axa-fr/react-toolkit-core';
import {
  CustomFormEvent,
  defaultOnChange,
} from '@axa-fr/react-toolkit-form-core';

export interface RadioItemProps
  extends Omit<InputHTMLAttributes<HTMLInputElement>, 'onChange'> {
  optionClassName?: string;
  inputRef?: React.LegacyRef<HTMLInputElement>;
  label: string;
  onChange?: (event: CustomFormEvent) => void;
  classModifier?: string;
}

const RadioItem: React.FC<RadioItemProps> = ({
  className = defaultClassName,
  disabled,
  value = '',
  id,
  inputRef,
  readOnly,
  onChange,
  children,
  label,
  classModifier = '',
  optionClassName,
  ...otherProps
}) => {
  const newLabel = children || label;
  const newId = InputManager.getInputId(id); // id is required on this component
  const classModifierWithDisabled = `${classModifier}${
    disabled ? ' disabled' : ''
  }`;
  const divOptionClassName = ClassManager.getComponentClassName(
    className,
    classModifierWithDisabled,
    defaultClassName
  );
  return (
    <div className={`${divOptionClassName} ${classModifier}`}>
      <input
        className="af-form__input-radio"
        value={value}
        id={newId}
        type="radio"
        onChange={defaultOnChange(onChange)}
        ref={inputRef}
        disabled={disabled}
        {...otherProps}
      />
      <label className="af-form__label" htmlFor={newId}>
        <span className="af-form__description">{newLabel}</span>
      </label>
    </div>
  );
};

export const defaultClassName = 'af-form__radio';

export default RadioItem;
