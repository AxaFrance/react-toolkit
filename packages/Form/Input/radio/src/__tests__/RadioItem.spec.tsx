import React from 'react';
import { render } from '@testing-library/react';
import RadioItem from '../RadioItem';

describe('RadioItem', () => {
  it('should have correct html rendered', () => {
    const { asFragment } = render(<RadioItem id="id" label="test" />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('input', () => {
    it('should call getInputId and set the input id', () => {
      const { getByRole } = render(<RadioItem id="id" label="test" />);
      expect(getByRole('radio')).toHaveAttribute('id', 'id');
    });
  });

  describe('div wrapper', () => {
    it('should set className to $expected when ', () => {
      const className = 'className';
      const { container } = render(
        <RadioItem id="id" label="test" className={className} />
      );
      expect(container.firstChild).toHaveClass(className);
    });

    it('should call getComponentClassName with disabled modifier added to classModifier', () => {
      const className = 'className';
      const classModifier = 'classModifier';

      const { container } = render(
        <RadioItem
          id="id"
          label="test"
          className={className}
          classModifier={classModifier}
          disabled
        />
      );

      expect(container.firstChild).toHaveClass(className);
      expect(container.firstChild).toHaveClass(
        `${className}--${classModifier}`
      );
    });
  });
});
