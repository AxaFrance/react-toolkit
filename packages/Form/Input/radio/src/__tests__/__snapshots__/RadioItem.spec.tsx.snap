// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioItem should have correct html rendered 1`] = `
<DocumentFragment>
  <div
    class="af-form__radio"
  >
    <input
      class="af-form__input-radio"
      id="id"
      type="radio"
      value=""
    />
    <label
      class="af-form__label"
      for="id"
    >
      <span
        class="af-form__description"
      >
        test
      </span>
    </label>
  </div>
</DocumentFragment>
`;
